`TestIntegrations/CmdLabels` flakiness

## Failure

#### Link(s) to logs

- https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695

#### Relevant snippet

```
integration_test.go:5237: 
[976](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:977)
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:5237
[977](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:978)
        	Error:      	Received unexpected error:
[978](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:979)
        	            	failed connecting to node 00000000-0000-0000-0000-000000000000. Teleport proxy failed to connect to "node" agent "" over reverse tunnel:
[979](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:980)
        	            	
[980](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:981)
        	            	  no tunnel connection found: no node reverse tunnel for 00000000-0000-0000-0000-000000000000.local-site found
[981](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:982)
        	            	
[982](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:983)
        	            	This usually means that the agent is offline or has disconnected. Check the
[983](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:984)
        	            	agent logs and, if the issue persists, try restarting it or re-registering it
[984](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:985)
        	            	with the cluster.
[985](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:986)
        	Test:       	TestIntegrations/CmdLabels/Worker_only
[986](https://github.com/gravitational/teleport/actions/runs/4067295252/jobs/7004473695#step:6:987)
        --- FAIL: TestIntegrations/CmdLabels/Worker_only (1.14s)```

Different failure on branch/v15 https://github.com/gravitational/teleport/actions/runs/8913671313/job/24479656198

```
=== FAIL: integration TestIntegrations/CmdLabels/Both (1.91s)
WARNING: Multiple nodes matched label selector, running command on all.
Running command on server-02:
Running command on server-01:

[server-02] success
[server-01] success

2 host(s) succeeded; 0 host(s) failed
    integration_test.go:6315: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:6315
        	Error:      	[]string{"[server-01] two[server-02] ", "two"} does not contain "[server-01] two"
        	Test:       	TestIntegrations/CmdLabels/Both
        --- FAIL: TestIntegrations/CmdLabels/Both (1.91s)
```

v16 https://github.com/gravitational/teleport/actions/runs/9601609534/job/26480573055#step:6:1446

```
     integration_test.go:6252: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:6252
        	Error:      	"[[server-02] two]" should have 2 item(s), but has 1
        	Test:       	TestIntegrations/CmdLabels/Both
        --- FAIL: TestIntegrations/CmdLabels/Both (1.43s)

=== FAIL: integration TestIntegrations/CmdLabels/Worker_only (1.13s)
    integration_test.go:6250: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:6250
        	Error:      	Received unexpected error:
        	            	failed connecting to host 00000000-0000-0000-0000-000000000000:0: failed to receive cluster details response
        	            		failed to dial target host
        	            		Teleport proxy failed to connect to "node" agent "" over reverse tunnel:
        	            	
        	            	  no tunnel connection found: no node reverse tunnel for 00000000-0000-0000-0000-000000000000.local-site found
        	            	
        	            	This usually means that the agent is offline or has disconnected. Check the
        	            	agent logs and, if the issue persists, try restarting it or re-registering it
        	            	with the cluster.
        	Test:       	TestIntegrations/CmdLabels/Worker_only
        --- FAIL: TestIntegrations/CmdLabels/Worker_only (1.13s)
```

https://github.com/gravitational/teleport/actions/runs/9646182423/job/26602026666

```
=== FAIL: integration TestIntegrations/CmdLabels/Both (1.46s)
WARNING: Multiple nodes matched label selector, running command on all.
Running command on server-01:
Running command on server-02:

[server-01] success
[server-02] success

2 host(s) succeeded; 0 host(s) failed
    integration_test.go:6254: 
        	Error Trace:	/__w/teleport/teleport/integration/integration_test.go:6254
        	Error:      	[]string{"[server-02] two[server-01] ", "two"} does not contain "[server-01] two"
        	Test:       	TestIntegrations/CmdLabels/Both
        --- FAIL: TestIntegrations/CmdLabels/Both (1.46s)

=== FAIL: integration TestIntegrations/CmdLabels (5.56s)
```

--- a/integration/integration_test.go
+++ b/integration/integration_test.go
@@ -6076,7 +6076,7 @@ func testCmdLabels(t *testing.T, suite *integrationTestSuite) {
 	teleport := suite.NewTeleportWithConfig(t, nil, nil, makeConfig())
 	defer teleport.StopAll()
 
-	// Create and start a reversetunnel node.
+	// Create and start a node.
 	nodeConfig := func() *servicecfg.Config {
 		tconf := suite.defaultServiceConfig()
 		tconf.Hostname = "server-02"
@@ -6088,8 +6088,20 @@ func testCmdLabels(t *testing.T, suite *integrationTestSuite) {
 
 		return tconf
 	}
-	_, err := teleport.StartReverseTunnelNode(nodeConfig())
+	_, err := teleport.StartNode(nodeConfig())
 	require.NoError(t, err)
+
+	slowPrintCommand := func(s string) []string {
+		cmd := make([]string, 0)
+		for i, char := range strings.Split(s, "") {
+			if i != 0 {
+				cmd = append(cmd, "&&")
+			}
+			cmd = append(cmd, "echo", "-n", char, "&&", "sleep", "0.05")
+		}
+
+		return cmd
+	}
 
 	// test label patterns that match both nodes, and each
 	// node individually.
@@ -6100,10 +6112,11 @@ func testCmdLabels(t *testing.T, suite *integrationTestSuite) {
 		expectLines []string
 	}{
 		{
-			desc:        "Both",
-			command:     []string{"echo", "two"},
+			desc: "Both",
+			// Print slowly so we can confirm that the output isn't interleaved.
+			command:     slowPrintCommand("abcd1234"),
 			labels:      map[string]string{"spam": "eggs"},
-			expectLines: []string{"[server-01] two", "[server-02] two"},
+			expectLines: []string{"[server-01] abcd1234", "[server-02] abcd1234"},
 		},
 		{
 			desc:        "Worker only",

--- a/lib/client/api.go
+++ b/lib/client/api.go
@@ -47,6 +47,7 @@ import (
 	"golang.org/x/crypto/ssh/agent"
 	"golang.org/x/net/http2"
 	"golang.org/x/sync/errgroup"
+	"golang.org/x/term"
 	"google.golang.org/grpc"
 
 	"github.com/gravitational/teleport"
@@ -93,6 +94,7 @@ import (
 	"github.com/gravitational/teleport/lib/tlsca"
 	"github.com/gravitational/teleport/lib/utils"
 	"github.com/gravitational/teleport/lib/utils/agentconn"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
 )
 
 const (
@@ -2769,6 +2771,33 @@ func (tc *TeleportClient) runCommandOnNodes(ctx context.Context, clt *ClusterCli
 
 	g, gctx := errgroup.WithContext(ctx)
 	g.SetLimit(commandLimit(ctx, clt.AuthClient, mfaRequiredCheck.Required))
+
+	// Get the width of the terminal so we can wrap properly.
+	var width int
+	if stdoutFile, ok := tc.Stdout.(*os.File); ok {
+		conn, err := stdoutFile.SyscallConn()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		ctrlErr := conn.Control(func(fd uintptr) {
+			width, _, err = term.GetSize(int(fd))
+			// If stdout is not a terminal, continue with a zero width instead
+			// of failing.
+			if err != nil {
+				width = 0
+			}
+		})
+		if ctrlErr != nil {
+			return trace.Wrap(ctrlErr)
+		}
+	}
+
+	stdout := logutils.NewSharedWriter(tc.Stdout)
+	stderr := stdout
+	if tc.Stdout != tc.Stderr {
+		stderr = logutils.NewSharedWriter(tc.Stderr)
+	}
+
 	for _, node := range nodes {
 		node := node
 		g.Go(func() error {
@@ -2803,7 +2832,12 @@ func (tc *TeleportClient) runCommandOnNodes(ctx context.Context, clt *ClusterCli
 			displayName := nodeName(node)
 			fmt.Printf("Running command on %v:\n", displayName)
 
-			if err := nodeClient.RunCommand(ctx, command, WithLabeledOutput()); err != nil && tc.ExitStatus == 0 {
+			if err := nodeClient.RunCommand(
+				ctx,
+				command,
+				WithLabeledOutput(width),
+				WithOutput(stdout, stderr),
+			); err != nil && tc.ExitStatus == 0 {
 				fmt.Fprintln(tc.Stderr, err)
 				return nil
 			}

--- a/lib/client/client.go
+++ b/lib/client/client.go
@@ -422,68 +422,93 @@ func (c *NodeClient) RunInteractiveShell(ctx context.Context, mode types.Session
 
 // lineLabeledWriter is an io.Writer that prepends a label to each line it writes.
 type lineLabeledWriter struct {
-	linePrefix        []byte
-	w                 io.Writer
-	shouldWritePrefix bool
+	linePrefix    string
+	w             io.Writer
+	maxLineLength int
+	buf           *bytes.Buffer
 }
 
-func newLineLabeledWriter(w io.Writer, label string) io.Writer {
+const defaultLabeledLineLength = 1024
+
+func newLineLabeledWriter(w io.Writer, label string, maxLineLength int) (io.WriteCloser, error) {
+	prefix := "[" + label + "] "
+	if maxLineLength == 0 {
+		maxLineLength = defaultLabeledLineLength
+	}
+	if maxLineLength <= len(prefix) {
+		return nil, trace.BadParameter("maxLineLength of %v is too short", maxLineLength)
+	}
+
+	buf := &bytes.Buffer{}
+	buf.Grow(maxLineLength + 1)
 	return &lineLabeledWriter{
-		linePrefix:        []byte(fmt.Sprintf("[%v] ", label)),
-		w:                 w,
-		shouldWritePrefix: true,
-	}
+		linePrefix:    prefix,
+		w:             w,
+		maxLineLength: maxLineLength,
+		buf:           buf,
+	}, nil
 }
 
-func (lw *lineLabeledWriter) writeChunk(b []byte, bytesWritten int, newline bool) (int, error) {
-	n, err := lw.w.Write(b)
-	bytesWritten += n
-	if err != nil {
-		return bytesWritten, trace.Wrap(err)
-	}
-	if newline {
-		n, err = lw.w.Write([]byte("\n"))
-		bytesWritten += n
-	}
-	return bytesWritten, trace.Wrap(err)
-}
-
+// Write writes data to the output writer. The underlying writer will only
+// receive complete lines at a time.
 func (lw *lineLabeledWriter) Write(input []byte) (int, error) {
 	bytesWritten := 0
-	var line []byte
 	rest := input
-	var found bool
-	for {
-		line, rest, found = bytes.Cut(rest, []byte("\n"))
-		// Write the prefix unless we're either continuing a line from the last
-		// write or we're at the end.
-		if lw.shouldWritePrefix && (len(line) > 0 || found) {
-			// Write the prefix on its own to not mess with the eventual returned
-			// number of bytes written.
-			if _, err := lw.w.Write(lw.linePrefix); err != nil {
+
+	for len(rest) > 0 {
+		origLine := rest
+		var line []byte
+		var writeLine bool
+		line, rest, writeLine = bytes.Cut(origLine, []byte("\n"))
+		// If the buffer is empty and we receive new data, it's a new line and
+		// we should add the prefix.
+		if lw.buf.Len() == 0 && (len(line) > 0 || writeLine) {
+			lw.buf.WriteString(lw.linePrefix)
+		}
+
+		// If we overflowed a line, cut a little earlier.
+		if lw.buf.Len()+len(line) > lw.maxLineLength {
+			linePortion := lw.maxLineLength - lw.buf.Len()
+			line = origLine[:linePortion]
+			rest = origLine[linePortion:]
+			writeLine = true
+			// We inserted this newline, don't count it later.
+			bytesWritten--
+		}
+
+		lw.buf.Write(line)
+		bytesWritten += len(line)
+		// If we hit a newline (or overflowed into one), flush the buffer.
+		if writeLine {
+			lw.buf.WriteString("\n")
+			bytesWritten++
+			_, err := lw.buf.WriteTo(lw.w)
+			if err != nil {
 				return bytesWritten, trace.Wrap(err)
 			}
-		}
-		var err error
-		if bytesWritten, err = lw.writeChunk(line, bytesWritten, found); err != nil {
-			return bytesWritten, trace.Wrap(err)
-		}
-		lw.shouldWritePrefix = true
-
-		if !found {
-			// If there were leftovers, the line will continue on the next write, so
-			// skip the first prefix next time.
-			lw.shouldWritePrefix = len(line) == 0
-			break
 		}
 	}
 
 	return bytesWritten, nil
 }
 
+// Close flushes the rest of the buffer to the output writer.
+func (lw *lineLabeledWriter) Close() error {
+	if lw.buf.Len() == 0 {
+		return nil
+	}
+	// End whatever line we're on to prevent clobbering.
+	lw.buf.WriteString("\n")
+	_, err := lw.buf.WriteTo(lw.w)
+	return trace.Wrap(err)
+}
+
 // RunCommandOptions is a set of options for NodeClient.RunCommand.
 type RunCommandOptions struct {
-	labelLines bool
+	labelLines    bool
+	maxLineLength int
+	stdout        io.Writer
+	stderr        io.Writer
 }
 
 // RunCommandOption is a functional argument for NodeClient.RunCommand.
@@ -491,9 +516,19 @@ type RunCommandOption func(opts *RunCommandOptions)
 
 // WithLabeledOutput labels each line of output from a command with the node's
 // hostname.
-func WithLabeledOutput() RunCommandOption {
+func WithLabeledOutput(maxLineLength int) RunCommandOption {
 	return func(opts *RunCommandOptions) {
 		opts.labelLines = true
+		opts.maxLineLength = maxLineLength
+	}
+}
+
+// WithOutput sends command output to the given stdout and stderr instead of
+// the node client's.
+func WithOutput(stdout, stderr io.Writer) RunCommandOption {
+	return func(opts *RunCommandOptions) {
+		opts.stdout = stdout
+		opts.stderr = stderr
 	}
 }
 
@@ -506,18 +541,40 @@ func (c *NodeClient) RunCommand(ctx context.Context, command []string, opts ...R
 	)
 	defer span.End()
 
-	var options RunCommandOptions
+	options := RunCommandOptions{
+		stdout: c.TC.Stdout,
+		stderr: c.TC.Stderr,
+	}
 	for _, opt := range opts {
 		opt(&options)
 	}
 
 	// Set up output streams
-	stdout := c.TC.Stdout
-	stderr := c.TC.Stderr
+	stdout := options.stdout
+	stderr := options.stderr
 	if c.hostname != "" {
 		if options.labelLines {
-			stdout = newLineLabeledWriter(c.TC.Stdout, c.hostname)
-			stderr = newLineLabeledWriter(c.TC.Stderr, c.hostname)
+			var err error
+			stdoutWriter, err := newLineLabeledWriter(
+				options.stdout,
+				c.hostname,
+				options.maxLineLength,
+			)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			defer stdoutWriter.Close()
+			stdout = stdoutWriter
+			stderrWriter, err := newLineLabeledWriter(
+				options.stderr,
+				c.hostname,
+				options.maxLineLength,
+			)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			defer stderrWriter.Close()
+			stderr = stderrWriter
 		}
 
 		if c.sshLogDir != "" {

--- a/lib/client/client_test.go
+++ b/lib/client/client_test.go
@@ -304,14 +304,15 @@ func (l wrappedListener) Accept() (net.Conn, error) {
 func TestLineLabeledWriter(t *testing.T) {
 	t.Parallel()
 	tests := []struct {
-		name     string
-		inputs   []string
-		expected string
+		name       string
+		inputs     []string
+		lineLength int
+		expected   string
 	}{
 		{
 			name:     "typical input",
 			inputs:   []string{"this is\nsome test\ninput"},
-			expected: "[label] this is\n[label] some test\n[label] input",
+			expected: "[label] this is\n[label] some test\n[label] input\n",
 		},
 		{
 			name:     "don't add empty line at end",
@@ -321,7 +322,7 @@ func TestLineLabeledWriter(t *testing.T) {
 		{
 			name:     "blank lines in middle",
 			inputs:   []string{"this\n\nis\n\nsome input"},
-			expected: "[label] this\n[label] \n[label] is\n[label] \n[label] some input",
+			expected: "[label] this\n[label] \n[label] is\n[label] \n[label] some input\n",
 		},
 		{
 			name:     "line break between writes",
@@ -331,18 +332,31 @@ func TestLineLabeledWriter(t *testing.T) {
 		{
 			name:     "line break immediately on second write",
 			inputs:   []string{"line 1", "\nline 2"},
-			expected: "[label] line 1\n[label] line 2",
+			expected: "[label] line 1\n[label] line 2\n",
 		},
 		{
 			name:     "line continues between writes",
-			inputs:   []string{"this is all ", "one continuous line"},
-			expected: "[label] this is all one continuous line",
+			inputs:   []string{"this is all ", "one continuous line ", "until\nnow"},
+			expected: "[label] this is all one continuous line until\n[label] now\n",
+		},
+		{
+			name:       "long lines wrapped",
+			inputs:     []string{"1234\nabcdefghijklmnopqrstuvwxyz\n1234"},
+			lineLength: 16,
+			expected:   "[label] 1234\n[label] abcdefgh\n[label] ijklmnop\n[label] qrstuvwx\n[label] yz\n[label] 1234\n",
+		},
+		{
+			name:       "exact length lines",
+			inputs:     []string{"abcdefgh", "\nijklmnop"},
+			lineLength: 16,
+			expected:   "[label] abcdefgh\n[label] ijklmnop\n",
 		},
 	}
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
 			var buf bytes.Buffer
-			w := newLineLabeledWriter(&buf, "label")
+			w, err := newLineLabeledWriter(&buf, "label", tc.lineLength)
+			require.NoError(t, err)
 
 			totalBytes := 0
 			expectedBytes := 0
@@ -352,6 +366,7 @@ func TestLineLabeledWriter(t *testing.T) {
 				totalBytes += n
 				expectedBytes += len(line)
 			}
+			assert.NoError(t, w.Close())
 			assert.Equal(t, expectedBytes, totalBytes)
 			assert.Equal(t, tc.expected, buf.String())
 		})

